subsys/net/lib/aws_fota/src/aws_fota.c

```diff
@@ -0,0 +1,507 @@
+		ERR_CHECK(err);
+
+		err = construct_job_id_update_topic(client->client_id.utf8,
+					      job_id,
+					      "/accepted",
+					      job_id_update_accepted_topic);
+		ERR_CHECK(err);
+		err = construct_job_id_update_topic(client->client_id.utf8,
+					      job_id,
+					      "/rejected",
+					      job_id_update_rejected_topic);
+
+		fota_state = DOWNLOAD_FIRMWARE;
+		update_job_execution(client,
+				     job_id,
+				     IN_PROGRESS,
+				     fota_state,
+				     doc_version_number,
+		 /* Client token are not used by this library */
+				     "");
+
+		/* Handled by the library*/
+		return 1;
+
+	} else if (!strncmp(job_id_update_accepted_topic,
+			    topic,
+			    MIN(JOB_ID_UPDATE_TOPIC_MAX_LEN, topic_len))) {
+		//err = parse_accepted_topic_payload(json_payload, status,
+		//&doc_version_number);
+		/* Set state to IN_PROGRES */
+		//execution_state = status;
+		if (execution_state == IN_PROGRESS &&
+		    fota_state == DOWNLOAD_FIRMWARE) {
+			fota_download_start(hostname, file_path);
+		}
+		else if (execution_state == IN_PROGRESS &&
+		    fota_state == APPLY_FIRMWARE) {
+			//clean up and report status
+			fota_state = NONE; //Maybe keep applying firmware as we haven't rebooted
+			update_job_execution(client,
+					     job_id,
+					     SUCCEEDED,
+					     fota_state,
+					     doc_version_number,
+					     "");
+		}
+		else if (execution_state == SUCCEEDED &&
+		    fota_state == APPLY_FIRMWARE) {
+			//callback_emit(AWS_FOTA_EVT_FINISHED);
+		}
+	}
+	return 0;
+
+}
+
+/**
+ * Return 0 for events which should be further processed by the client
+ * Return >= 1 for events which can be skipped/events handled by the
+ * library.
+ */
+int aws_fota_mqtt_evt_handler(struct mqtt_client *const client,
+			      const struct mqtt_evt *evt)
+{
+	int err;
+	switch (evt->type) {
+	case MQTT_EVT_CONNACK:
+		if (evt->result != 0) {
+			return 0;
+		}
+
+		err = construct_notify_next_topic(client->client_id.utf8,
+						  notify_next_topic);
+		if (err) {
+			return err;
+		}
+
+		err = aws_jobs_subscribe_expected_topics(client, true);
+		if (err) {
+			return err;
+		}
+
+		err = update_device_shadow_version(client);
+		if (err) {
+			return err;
+		}
+
+		return 0;
+		/* This expects that the application's mqtt handler will handle
+		 * any situations where you could not connect to the MQTT
+		 * broker.
+		 */
+
+	case MQTT_EVT_DISCONNECT:
+		return 0;
+
+	case MQTT_EVT_PUBLISH: {
+		const struct mqtt_publish_param *p = &evt->param.publish;
+		//get_payload
+		err = publish_get_payload(client, p->message.payload.len);
+
+		err = aws_fota_on_publish_evt(client,
+					      p->message.topic.topic.utf8,
+					      p->message.topic.topic.size,
+					      payload_buf);
+
+	} break;
+
+	case MQTT_EVT_PUBACK:
+		if (evt->result != 0) {
+			LOG_ERR("MQTT PUBACK error %d\n", evt->result);
+			return 0;
+		}
+		/* check evt->param.puback.message_id */
+		/* We expect that the client pubacks*/
```

Comment:
If you receive QoS1, you should puback on MQTT_EVT_PUBLISH, see https://github.com/zephyrproject-rtos/zephyr/blob/21358baa72cea9c23be57444eb91444774842f97/samples/net/google_iot_mqtt/src/protocol.c#L164

Reply:
So the idea here since we are hijacking the evt is for the application to do puback upon QoS1 evt. but this would maybe conflict with the idea of giving the application the possibility of skip handling. I'll look a bit into it.

Reply:
Well,  I think that if we pursue the idea, that some events are handled by the `aws_fota` lib, they should be handled entirely, so that app can immediately leave the event handler once it sees that the event got already handled.

Reply:
Yeah. I'll do that :+1: